File: frontend/lib/proc/quote.dart
class Quote {
  final String? id;
  final String procurementId;
  final String supplierId;
  final List<QuoteItem> items;
  final double totalCost;

  Quote({
    this.id,
    required this.procurementId,
    required this.supplierId,
    required this.items,
    required this.totalCost,
  });

  factory Quote.fromJson(Map<String, dynamic> json) {
    var itemsJson = json['items'] as List;
    List<QuoteItem> items = itemsJson.map((i) => QuoteItem.fromJson(i)).toList();

    return Quote(
      id: json['_id'],
      procurementId: json['procurement_id'],
      supplierId: json['supplier_id'],
      items: items,
      totalCost: json['total_cost'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'procurement_id': procurementId,
      'supplier_id': supplierId,
      'items': items.map((item) => item.toJson()).toList(),
      'total_cost': totalCost,
    };
  }
}

class QuoteItem {
  final String itemId;
  final int quantity;
  final double unitPrice;

  QuoteItem({
    required this.itemId,
    required this.quantity,
    required this.unitPrice,
  });

  factory QuoteItem.fromJson(Map<String, dynamic> json) {
    return QuoteItem(
      itemId: json['item_id'],
      quantity: json['quantity'],
      unitPrice: json['unit_price'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'item_id': itemId,
      'quantity': quantity,
      'unit_price': unitPrice,
    };
  }
}

File: backend/quote.go
package main

import (
	"context"
	"encoding/json"
	"net/http"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Quote struct {
	ID            primitive.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	ProcurementID primitive.ObjectID `bson:"procurement_id" json:"procurement_id"`
	SupplierID    primitive.ObjectID `bson:"supplier_id" json:"supplier_id"`
	Items         []QuoteItem        `bson:"items" json:"items"`
	TotalCost     float64            `bson:"total_cost" json:"total_cost"`
}

type QuoteItem struct {
	ItemID    primitive.ObjectID `bson:"item_id" json:"item_id"`
	Quantity  int                `bson:"quantity" json:"quantity"`
	UnitPrice float64            `bson:"unit_price" json:"unit_price"`
}

func getQuotes(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	var quotes []Quote
	collection := client.Database("procurementdb").Collection("quotes")
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	cursor, err := collection.Find(ctx, bson.M{})
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		w.Write([]byte(`{ "message": "` + err.Error() + `" }`))
		return
	}
	defer cursor.Close(ctx)
	for cursor.Next(ctx) {
		var quote Quote
		cursor.Decode(&quote)
		quotes = append(quotes, quote)
	}
	if err := cursor.Err(); err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		w.Write([]byte(`{ "message": "` + err.Error() + `" }`))
		return
	}
	json.NewEncoder(w).Encode(quotes)
}

func createQuote(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	var quote Quote
	json.NewDecoder(r.Body).Decode(&quote)
	collection := client.Database("procurementdb").Collection("quotes")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	result, err := collection.InsertOne(ctx, quote)
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		w.Write([]byte(`{ "message": "` + err.Error() + `" }`))
		return
	}
	json.NewEncoder(w).Encode(result)
}
